nibabel/streamlines/tck.py

```diff
@@ -401,46 +396,36 @@ def _read(cls, fileobj, header, buffer_size=4):
             f.seek(header["_offset_data"], os.SEEK_SET)
 
             eof = False
-            buffs = []
             n_streams = 0
-
+            leftover = np.empty([0,3])
             while not eof:
-                buff = bytearray(buffer_size)
-                n_read = f.readinto(buff)
-                eof = n_read != buffer_size
-                if eof:
-                    buff = buff[:n_read]
-
-                buffs.append(buff)
-
-                # Make sure we've read enough to find a streamline delimiter.
-                if fiber_marker not in buff:
-                    # If we've read the whole file, then fail.
-                    if eof:
-                        # Could have minimal buffering, and have read only the
-                        # EOF delimiter
-                        buffs = [bytearray().join(buffs)]
-                        if not buffs[0] == eof_marker:
-                            raise DataError(
-                                "Cannot find a streamline delimiter. This file"
-                                " might be corrupted.")
-                    else:
-                        # Otherwise read a bit more.
-                        continue
-
-                all_parts = bytearray().join(buffs).split(fiber_marker)
-                point_parts, buffs = all_parts[:-1], all_parts[-1:]
-                point_parts = [p for p in point_parts if p != b'']
-
-                for point_part in point_parts:
-                    # Read floats.
-                    pts = np.frombuffer(point_part, dtype=dtype)
-                    # Convert data to little-endian if needed.
-                    yield pts.astype('<f4', copy=False).reshape([-1, 3])
-
-                n_streams += len(point_parts)
-
-            if not buffs[-1] == eof_marker:
+
+                # read raw files from file
+                raw_values = np.fromfile(f.fobj, dtype, batch_size)
+                if len(raw_values) < batch_size:
+                    eof = True
+
+                # Convert raw_values into a list of little-endian tuples (for x,y,z coord)
+                coords = raw_values.astype('<f4', copy=False).reshape([-1, 3])
+
+                # find stream delimiter locations (all NaNs)
+                delims = np.where(np.all(np.isnan(coords), axis=1))[0]
+
+                # for each delimiters, yeild new streams
+                begin = 0
+                for i in range(0, len(delims)):
+                    end = delims[i]
+                    stream = np.append(leftover, coords[begin:end], axis=0)
+                    leftover = np.empty([0,3])
+                    yield stream
+                    n_streams += 1
+
+                    begin = end+1 #skip the delimiter
+
+                # the rest gets appended to the leftover
+                leftover = np.append(leftover, coords[begin:], axis=0)
```

Comment:
I don't know if this is going to be a major contributor to the slowdown, but we can reduce the number of appends to <=1 per batch instead of 1 per streamline + 1 per batch:

```Python
delims = np.where(np.all(np.isnan(coords), axis=1))[0]

if leftover.size:
    coords = np.vstack((leftover, coords))
    delims += leftover.shape[0]
    leftover = np.empty((0, 3), dtype='<f4')

begin = 0
for delim in delims:
    yield coords[begin:delim]
    n_streams += 1
    begin = delim + 1

leftover = coords[begin:]